pkg/piecestore/rpc/client/readerwriter.go

```diff
@@ -138,35 +129,18 @@ func NewStreamReader(signer *Client, stream pb.PieceStoreRoutes_RetrieveClient,
 				break
 			}
 
-			if sr.allocated-sr.totalRead >= int64(4*signer.bandwidthMsgSize) || sr.max-sr.allocated < int64(4*signer.bandwidthMsgSize) {
-				fmt.Println("Sending Signal to dataSendCond")
-				time.Sleep(time.Second * 10)
-				sr.dataSendCond.Signal() // Signal the data send loop that it has available bandwidth
-				fmt.Println("Sent Signal to dataSendCond")
-			}
 		}
 	}()
 
 	sr.src = utils.NewReaderSource(func() ([]byte, error) {
-		sr.dataSendCond.L.Lock()
-		fmt.Println("Locking data send")
-		for sr.totalRead <= sr.allocated {
-			sr.c <- 1
-			fmt.Println("dataSendCond Wait")
-			sr.dataSendCond.Wait() // Wait for the bandwidth loop to send a new bandwidth allocation
-		}
-		fmt.Println("Unlocking data send")
-		sr.dataSendCond.L.Unlock()
-
+		_, err := sr.throttle.ConsumeOrWait(int64(signer.bandwidthMsgSize))
```

Comment:
There's no need to block while the sender is being generous. In general the part that receives messages should never block.